### ISSUE TITLE
Feature Request: Specify Ordering of Help Output

### ISSUE THREAD

**Issue by sirosen** [pallets/click#672]
I'm happy to put together a PR that does this, but I want agreement on syntax/design and desirability of this feature before putting in the effort.

At present, `click.help_option()` makes its own decisions about the order of options and subcommands in the help output.
When you have a command tree that's variegated, it becomes hard to present this information cleanly.
Consider these commands (groups implied):

``` shell
$ foo alpha beta gamma
$ foo alpha phi eta
$ foo alpha delta
$ foo alpha mu
$ foo tau psi
$ foo epsilon
```

If you run with default help, you could get an ordering like this:

``` shell
$ foo --help
...  # options
Commands:
  alpha
  epsilon
  tau

$ foo alpha --help
...  # options
Commands:
  mu
  beta
  phi
  delta
```

The important thing is that groups and commands are being interspersed arbitrarily in the Commands section of the help output.
This is much worse than putting all groups first or all groups last in terms of readability for a new user.

The same basic issue applies for options -- if you want all of the required options first or last, or all of the options with defaults last, or whatever other ordering, you're at the mercy of the default help option.

This situation can be significantly improved by letting users specify an order for the help output.
I see a couple of solutions for this.

The help opt could take an ordering callback. Example:
`command_sort=None`, `parameter_sort=None`, optional pair of callables that determine the order of commands and parameters. Consume an iterable of `click.Command` objects and/or an iterable of `click.Parameter` objects and return them in sorted order. Defaults to alphabetical order.

``` python
# put MultiCommands after all other commands -- no other enforced ordering
def command_sort(commands):
    return sorted(commands, key=lambda x: isinstance(x, click.MultiCommand))
# command is `f`
f = click.help_option(command_sort=command_sort)(f)
```

Or each option and command could have a "help precedence" integer value that determines sort order and defaults to 0.
Then people can tune the value up and down around 0 and the help option will know to respect those values.
Example:

``` python
# guarantees that options will appear in order as
# --foo, --baz, --bar
@click.option('--foo', help_precedence=-1)
@click.option('--bar', help_precedence=1)
@click.option('--baz')
def cmdfunc(...):
    ...
```

The ordering functions are more extensible and probably a lot easier to implement, but harder to apply from a usage standpoint as "little tweaks" to the help output.

**Comment by chrisdl** [comment]
👍

**Comment by chrisdl** [comment]
I dug into this and a workable solution for now is as follows:

```py
class MyGroup(click.Group):
    def list_commands(self, ctx):
        """Override"""
        # original value --> return sorted(self.commands)
        return ['b', 'a', 'c']

@click.group(cls=MyGroup)
def main():
    pass

@main.command()
def a():
    pass

@main.command()
def b():
    pass
```

I am fairly certain this is achievable by overriding without using the group as well, I see that CommandCollection has a list_commands function as well.

see this for what we are overriding: https://github.com/pallets/click/blob/67f39e4ff9093ee347207d1cc1abe0130c21268e/click/core.py#L1188

Hope that helps.

**Comment by vlcinsky** [comment]
Very similar to #513, consider merging it.

One natural solution for defining order would be the order it appears in the code. I am not sure, it this is easy to implement, it could be either trivial (the sorting would return items as they come in), or difficult (if it would be difficult to detect the order of items according to the code).

**Comment by sirosen** [comment]
@chrisdl I'm glad you mentioned that the original implementation is using `sorted(self.commands)`.
Interestingly, it doesn't look like any of the `click.Command` subclasses define comparators, so that sorting gives a stable ordering, but not a meaningful one.

That suggests another good solution: allow users to define the behavior of `click.Command.__cmp__`, and inherit `__cmp__` from groups when attaching a command unless it is already specified.
So, I should be able to define `__cmp__` once at the root of my command heirarchy, but also override it on any subcommands that have other desirable orderings.

@vlcinsky You seem to be asking for Click to keep track of the order in which commands are added, and use that as the sort ordering. It's not a terrible solution (or, IMO, even a particularly hard one), and probably a sane enough default, but I don't like the connection it creates between implementation details and the UX. Moving pieces of code around without changing their functionality shouldn't change their behavior -- or if it does, we need to be able to specify that behavior explicitly instead.

**Comment by vlcinsky** [comment]
@sirosen I usually keep the UI related code in single file importing other stuff from other modules and have no need to reshuffle order of UI (CLI) related functions around.

If it is easy to implement it that way (by default keep the order of commands in order they are added), I think, it would provide very intuitive result. In fact it would be more intuitive than current alphabetical sorting. Adding custom __cmp__ or other methods of sort definition is often tricky (at the moment one adds another subcommand and forgets to update the sort order in another function).

**Comment by vlcinsky** [comment]
In #513 I have described solution based on example from @srisdl which has resolved my need to see subcommands in the order I have defined them in my source code.

If such an implementation (commands as OrderedDict, list_commands not sorting anything) would become part of click solution, it would be good base for defining any other sorting strategy as this issue requests.

For alternative sorting strategies I would expect user to inherit from NaturalOrderGroup (or click.Group, if the NaturalOrderGroup approach would be accepted as base one) and redefined `list_commands`.

In `list_commands` one would have couple of options:
- define fixed order of subcommand names (as in example above)
- internally (in `list_commands` scope) use tuple `(start_names, tail_names)`. `start_names` would be list of subcommand names to come in given order in start of subcommand names. Only names in real self.commands would be returned. Similarly `tail_names` would define order of subcommand names to come to the end. Names in self.commands which are not present in either of `start_names` or `tail_names` would come into middle of returned list.

In fact this is sort of "define command `__cmp__`" but it is better located in place (a group, not command class itself) where it is relevant.
